docs/client-concepts/high-level/indexing/pipelines.asciidoc

```diff
@@ -0,0 +1,134 @@
+}
+
+public class GeoIp
+{
+    public string CityName { get; set; }
+    public string ContinentName { get; set; }
+    public string CountryIsoCode { get; set; }
+    public GeoLocation Location { get; set; }
+    public string RegionName { get; set; }
+}
+----
+
+==== Create an ingestion pipeline
+
+Assuming we are indexing Person documents, we can create an ingestion pipeline that manipulates the
+incoming values before they are indexed.
+
+Lets assume that our application always expects surnames to be capitalised, and for initials to
+be indexed into their own field. We also have an IP address that we'd like to convert into
+a human-readable location.
+
+We could achieve this requirement by creating a custom mapping and creating an ingest pipeline.
+The Person type can then be used as-is, without making any changes.
+
+[source,csharp]
+----
+client.CreateIndex("people");
+
+client.Map<Person>(p => p
+    .Index("people")
+    .AutoMap() <1>
+    .Properties(props => props
+        .Keyword(t => t.Name("initials")) <2>
+        .Ip(t => t.Name(dv => dv.IpAddress)) <3>
+        .Object<GeoIp>(t => t.Name(dv => dv.GeoIp)) <4>
+    )
+);
+
+client.PutPipeline("person-pipeline", p => p
+    .Processors(ps => ps
+        .Uppercase<Person>(s => s
+            .Field(t => t.LastName) <5>
+        )
+        .Script(s => s
+            .Lang("painless") <6>
+            .Source("ctx.initials = ctx.firstName.substring(0,1) + ctx.lastName.substring(0,1)")
+        )
+        .GeoIp<GeoIp>(s => s <7>
+            .Field("ipAddress")
+            .TargetField("geoIp")
+        )
+    )
+);
+
+var person = new Person
+{
+    Id = 1,
+    FirstName = "Martijn",
+    LastName = "Laarman",
+    IpAddress = "139.130.4.5"
+};
+
+var indexResponse = client.Index(person, p => p.Index("people").Pipeline("person-pipeline")); <8>
+----
+<1> automatically create the mapping from the type
+
+<2> create an additional field to store the initials
+
+<3> map field as IP Address type
+
+<4> map GeoIp as object
+
+<5> uppercase the lastname
+
+<6> use a painless script to populate the new field
+
+<7> use GeoUp plugin to enrich the GeoIp object from the supplied IP Address
+
+<8> index the document using the created pipeline
+
+==== Increasing timeouts
+
+When a pipeline is specified, there will be the added overhead of document enrichment when indexing, the example given above, the execution
+of the uppercasing and the painless script.
+
+For large bulk requests, it could be prudent to increase the default indexing timeout to avoid exceptions.
+
+[source,csharp]
+----
+client.Bulk(b => b
+    .Index("people")
+    .Pipeline("person-pipeline")
+    .Timeout("5m") <1>
+    .Index<Person>(/*snip*/)
+    .Index<Person>(/*snip*/)
+    .Index<Person>(/*snip*/)
+    .RequestConfiguration(rc => rc
+            .RequestTimeout(TimeSpan.FromMinutes(5)) <2>
+    )
+);
+----
+<1> increases the bulk timeout to 5 minutes
+
+<2> increases the request timeout to 5 minutes
```

Comment:
I think the difference between these two timeouts should be explained.

Reply:
Done.